src/MultiCollateralVault.sol

```diff
@@ -0,0 +1,275 @@
+        require(token_ != IERC20(address(0)), "token cannot be zero address");
+        require(rad > 0, "ratio must be greater than 0");
+        require(rad <= 1e27, "ratio must be less than or equal to 1");
+        require(collateralInfos[token_].oracle != AggregatorV3Interface(address(0)), "token has not been added as collateral");
+
+        collateralInfos[token_].ratio = rad;
+    }
+
+    
+    /// @notice allows authorized user to liquidate under collateralized Users
+    function liquidate(address user) external auth {
+        require(!_isCollateralized(user), "specified user can't be liquidated");
+         debt[user] = 0;
+        IERC20[] memory _collaterals = Collaterals;
+        uint256 _length = _collaterals.length;
+        for(uint256 i = 0; i < _length;) {
+            uint256 amount =  deposits[user][_collaterals[i]];
+            deposits[user][_collaterals[i]] = 0;
+            _collaterals[i].safeTransfer(msg.sender, amount);
+            unchecked{i++;}
+        }
+    } 
+
+/***
+-------------------------------------------------------------------------------------------------------------------------------------------------------------
+    Pricing/Vault Ratio functions
+-------------------------------------------------------------------------------------------------------------------------------------------------------------
+ */
+
+    /// @return price the value of token denominated in WETH
+    /// @param token address for the token that's being inquired about
+    function price(IERC20 token) public view returns(uint256) {
+        return _price(token);
+    }
+
+    function freeCollateral(address user, IERC20 token) view external returns(uint256) {
+        return _freeCollateral(user, token);
+    }
+
+    /// @dev used in testing
+    function remainingPower(address user) public view returns(uint256){
+        return _remainingPower(user);
+    }
+    
+    function _price(IERC20 token) internal view returns(uint256) {
+        AggregatorV3Interface iOracle = collateralInfos[token].oracle;
+        (   uint80 roundID,
+            int256 price_,
+            /*uint startedAt*/,
+            uint timeStamp,
+            uint80 answeredInRound
+        ) = iOracle.latestRoundData();
+        require(price_ > 0, "ChainLink: price <= 0");
+        require(answeredInRound >= roundID, "ChainLink: Stale price");
+        require(timeStamp > 0, "ChainLink: Round not complete");
+        return uint256(price_);
+    }
+
+    /// @dev calculates the total borrowing power of a users deposits in ETH, and the total value of a users in Deposits in ETH
+    function _borrowingPower(address user) internal view returns(uint256 borrowingPower_, uint256 depositsInETH_) {
+        IERC20[] memory _collaterals = Collaterals;
+        uint256 _length = _collaterals.length;
+        uint256 price_;
+        uint256 deposit_;
+        uint256 valueInETH;
+        for(uint256 i = 0; i < _length; ) {
```

Comment:
At the beginning of each loop, you should check if the user has any of the selected collateral, and skip to the next collateral if it doesn't, you do it for `remainingPower`.